### PR TITLE
docs: CONTRIBUTING.md

### LINKED ISSUE #178: Formally define the difference between `error`, `warning`, and `note` diagnostics
* `error` is pretty straightforward, it is used when parsed WDL is incorrect or invalid in some way.
* `warning` should be used when encountered WDL is confusing, problematic, error-prone, etc. but not "invalid".
* `note` would be everything else and is mostly issues of style conformity

If we all agree on the above, we should
1. write it down somewhere public that it can be referenced in the future
2. go through all our existing lints and make sure they use `warning`/`note` appropriately

### PR DESCRIPTION
**Pull request by a-frantz**
I opted for an FAQ format. Maybe the file should be called `FAQ.md` instead? But if we make that change, it should probably broaden in scope as all the current questions are very much focused on the development side of things.

Closes #178 and #181

[conventional commit]: https://www.conventionalcommits.org/en/v1.0.0/#summary
["keep a changelog"]: https://keepachangelog.com/en/1.0.0/

### CODE REVIEW

File: DEVELOPMENT.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+## What's a good first issue?
+
+We will try to keep a handful of [issues](https://github.com/stjude-rust-labs/wdl/issues) marked `good first issue` open and ready for new contributors.
+
+## I don't want to write code, can I still contribute?
+
+Sure!
+
+You can always open a [discussion](https://github.com/stjude-rust-labs/wdl/discussions/categories/rule-proposals) with a propsal for a new lint rule.
```
Comment (adthrasher) [inline DEVELOPMENT.md:19]:
```suggestion
You can always open a [discussion](https://github.com/stjude-rust-labs/wdl/discussions/categories/rule-proposals) with a propsal for a new lint rule. Or contribute to any open discussions.
```

File: DEVELOPMENT.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+We also appreciate feedback on our documentation. Feel free to look over any of our `*.md` files and note any issues you find. You can also explore our lint rule documentation by [installing `sprocket`](https://stjude-rust-labs.github.io/sprocket/installation.html) and reading the output of `sprocket explain`. (nota bene: we hope to replace `sprocket explain` with a website where each rule will have a dedicated page, but that has not been realized yet)
```
Comment (adthrasher) [inline DEVELOPMENT.md:21]: I'm not sure I've ever seen anyone write out _nota bene_ before. Maybe just use the abbreviation _n.b._.

File: DEVELOPMENT.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+## What's the difference between `error`, `warning`, and `note`?
+
+- an `error` is emitted when WDL is incorrect or invalid in some way
+- a `warning` is emitted when WDL is confusing, problematic, error-prone, etc. but not invalid or incorrect
```
Comment (adthrasher) [inline DEVELOPMENT.md:26]:
```suggestion
- an `error` is emitted when the source WDL is incorrect or invalid in some way
- a `warning` is emitted when the source WDL is confusing, problematic, error-prone, etc. but not invalid or incorrect
```

File: DEVELOPMENT.md [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+In order to turn the Gauntlet CI green, try running `cargo run --release --bin gauntlet -- --refresh`. The `--refresh` flag will save any changes to the `Gauntlet.toml` file. This should then be committed and included in your PR.
+
+## What is arena?
+
+Arena is the alternate run mode of `gauntlet`. [`Arena.toml`](https://github.com/stjude-rust-labs/wdl/blob/main/Arena.toml) is very similar to `Gauntlet.toml`, except it has fewer repository entries and instead of analysis diagnostics it contains only lint diagnostics (which are not included in `Gauntlet.toml`).
+
+In order to turn the Arena CI green, try running `cargo run --release --bin gauntlet -- --arena --refresh`. The `--refresh` flag (in conjunction with the `--arena` flag) will save any changes to the `Arena.toml` file. This should then be committed and included in your PR.
```
Comment (adthrasher) [inline DEVELOPMENT.md:39]:
```suggestion
In order to turn the Arena CI green, run `cargo run --release --bin gauntlet -- --arena --refresh`. The `--refresh` flag (in conjunction with the `--arena` flag) will save any changes to the `Arena.toml` file. This should then be committed and included in your PR.
```

File: DEVELOPMENT.md [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+- an `error` is emitted when WDL is incorrect or invalid in some way
+- a `warning` is emitted when WDL is confusing, problematic, error-prone, etc. but not invalid or incorrect
+- a `note` is emitted in all other cases and is mostly used for issues of style or conformity
+
+## What is gauntlet?
+
+[Gauntlet](https://github.com/stjude-rust-labs/wdl/tree/main/gauntlet) is the main driver of our CI. Take a look at the file [`Gauntlet.toml`](https://github.com/stjude-rust-labs/wdl/blob/main/Gauntlet.toml). The entries at the top are all GitHub repositories of WDL code. The remaining entries are diagnostics emitted while analyzing those repositories. These should remain relatively static between PRs, and any change in emitted diagnostics should be reviewed carefully.
+
+In order to turn the Gauntlet CI green, try running `cargo run --release --bin gauntlet -- --refresh`. The `--refresh` flag will save any changes to the `Gauntlet.toml` file. This should then be committed and included in your PR.
```
Comment (adthrasher) [inline DEVELOPMENT.md:33]:
```suggestion
In order to turn the Gauntlet CI green, run `cargo run --release --bin gauntlet -- --refresh`. The `--refresh` flag will save any changes to the `Gauntlet.toml` file. This should then be committed and included in your PR.
```

**Review by adthrasher — COMMENTED** [review]
Maybe this should be part of a `CONTRIBUTING.md`?

File: DEVELOPMENT.md (hunk 2 above)
Comment (a-frantz) [reply DEVELOPMENT.md:21]: 😅 😅  my years taking Latin are showing

**Comment by a-frantz** [comment]
> Maybe this should be part of a `CONTRIBUTING.md`?

I think I like titling this `CONTRIBUTING.md` better than the current name. Going to make that change now.

**Comment by adthrasher** [comment]
> > Maybe this should be part of a `CONTRIBUTING.md`?
> 
> I think I like titling this `CONTRIBUTING.md` better than the current name. Going to make that change now.

In that case, I think we should include some additional information in this file. It's probably worth looking at some other examples to see what can fairly be expected in a _contributing_ file. I don't necessarily think that has to be in this PR, we could just merge what you have and work on enhancing the file later.

**Comment by a-frantz** [comment]
@adthrasher has convinced me that this `CONTRIBUTING.md` file should be fleshed out a bit more, so I'm going to noodle at it a bit more. I think at the very least we need some sort of "welcome" message at the top before the FAQs start. And there's probably other things worth adding as well 🤔

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,69 @@
+# Welcome to the `wdl` crates!
+
+Community contributions rock and we are psyched you're reading this document!
```
Comment (a-frantz) [inline CONTRIBUTING.md:3]: Welcoming feedback on this opening.

I think it's important for a contributing guide to be friendly and welcoming to new folks; after all, we're trying to encourage them to join our effort. So I went pretty informal with the opening message here. Maybe there's a different tone we want to strike?

What do yall think?

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+- `cargo clippy --all-features` and then fix any warnings emitted
+- `BLESS=1 cargo test --all-features` to "bless" any test changes
+    - Please review any changes this causes to make sure they seem right!
+- `cargo run --release --bin gauntlet -- --refresh`
+    - see the `What is gauntlet?` question for more information
+- `cargo run --release --bin gauntlet -- --refresh --arena`
+    - see the `What is arena?` question for more information
+- `rustup update` to update your local toolchains
+
+### What's the general workflow for writing a new lint rule?
```
Comment (a-frantz) [inline CONTRIBUTING.md:70]: Below is generally how I do it. I got asked this question by a new contributor over Slack, and this is part of my response to them. I think the test infrastructure is a stumbling block, and I hope this Q+A helps with explaining how the tests should be used.

File: CONTRIBUTING.md (hunk 6 above)
Comment (peterhuene) [reply CONTRIBUTING.md:3]: I'm personally fine with it; not sure what tone @claymcleod might want, though.

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+### What's a good first issue?
+
+We will try to keep a handful of [issues][issues] marked `good first issue` open and ready for new contributors.
+
+### I don't want to write code, can I still contribute?
+
+Sure!
+
+You can always open a [discussion](https://github.com/stjude-rust-labs/wdl/discussions/categories/rule-proposals) with a propsal for a new lint rule or contribute to any open discussions.
```
Comment (peterhuene) [inline CONTRIBUTING.md:33]:
```suggestion
You can always open a [discussion](https://github.com/stjude-rust-labs/wdl/discussions/categories/rule-proposals) with a proposal for a new lint rule or contribute to any open discussions.
```

**Review by peterhuene — COMMENTED** [review]
Other than the typo, looks great!

**Review by peterhuene — APPROVED** [review]

**Pull request merged**